### PR TITLE
Add dead-branch-prune pass after SSA pass

### PR DESCRIPTION
**Pull request by sklam**
Fixes https://github.com/numba/numba/issues/9706
Fixes https://github.com/numba/numba/issues/6541
Fixes https://github.com/numba/numba/issues/7482
Fixes https://github.com/numba/numba/issues/5661

Add an extra dead-branch-prune pass after SSA

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the patch. I think the effect of this PR is moving the somewhat experimental "SSA compliant" part of the branch pruning pass from behind the `hasattr/getattr/isinstance` guard into the main pipeline. This is probably ok given functions like `hash` now look up `__hash__` on the type, such that this pass has likely been exercised reasonably often in external code without reported issue. Are you happy with that?

File: numba/tests/test_remove_dead.py [hunk 1]
```diff
@@ -295,5 +295,66 @@ def define_pipelines(self):
         np.testing.assert_array_equal(test_res, py_res)
 
 
+class TestSSADeadBranchPrune(TestCase):
+    """
+    Test issues that required dead-branch-prune on SSA IR
+    """
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_remove_dead.py:301]: I rather just test that the use-cases work. We just don't want an regression that break these added features

File: numba/tests/test_remove_dead.py [hunk 2]
```diff
@@ -295,5 +295,66 @@ def define_pipelines(self):
...
+            else:
+                assert np.all((0 <= out) & (out < N))
+            out[:] = N
+            return out
+
+        expected = f(np.array([3, 1, 2]))
+        out = np.arange(3, dtype='i8')
+        got = f(np.array([3, 1, 2]), out=out)
+        self.assertIs(got, out)
+        self.assertPreciseEqual(got, expected)
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_remove_dead.py:337]: done in 5cc0be194abd0ba8a7cd279e7f81b32662879e5b

File: numba/tests/test_remove_dead.py [hunk 3]
```diff
@@ -295,5 +295,66 @@ def define_pipelines(self):
...
+        @njit
+        def foo_manual_ssa(x, y=None):
+            if y is not None:
+                return x + y
+            else:
+                # avoid changing type of `y`
+                y_ = x
+                return x + y_
+
+        self.assertEqual(foo(3, None), foo_manual_ssa(3, None))
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_remove_dead.py:320]: done in 5cc0be194abd0ba8a7cd279e7f81b32662879e5b

**Comment by sklam** [comment]
I just discovered there is un-committed code in my local branch: https://github.com/numba/numba/pull/9711/commits/a5ac54c13436b9eba3a80a76528fdd46bbb548be, which add another test for an issue.

**Pull request merged**